networks/rpc/websocket.go

```diff
@@ -24,77 +24,70 @@ import (
 	"encoding/json"
 	"fmt"
-	"net"
 	"net/http"
 	"net/url"
 	"os"
 	"strings"
+	"sync"
 	"sync/atomic"
 	"time"
 
-	"github.com/klaytn/klaytn/common"
-	"golang.org/x/net/websocket"
-	"gopkg.in/fatih/set.v0"
-
 	fastws "github.com/clevergo/websocket"
+	mapset "github.com/deckarep/golang-set"
+	"github.com/gorilla/websocket"
+	"github.com/klaytn/klaytn/common"
 	"github.com/valyala/fasthttp"
 )
 
-// websocketJSONCodec is a custom JSON codec with payload size enforcement and
-// special number parsing.
-var websocketJSONCodec = websocket.Codec{
-	// Marshal is the stock JSON marshaller used by the websocket library too.
-	Marshal: func(v interface{}) ([]byte, byte, error) {
-		msg, err := json.Marshal(v)
-		return msg, websocket.TextFrame, err
-	},
-	// Unmarshal is a specialized unmarshaller to properly convert numbers.
-	Unmarshal: func(msg []byte, payloadType byte, v interface{}) error {
-		dec := json.NewDecoder(bytes.NewReader(msg))
-		dec.UseNumber()
-
-		return dec.Decode(v)
-	},
+const (
+	wsReadBuffer  = 1024
+	wsWriteBuffer = 1024
+)
+
+var wsBufferPool = new(sync.Pool)
+
+func newWebsocketCodec(conn *websocket.Conn) ServerCodec {
+	conn.SetReadLimit(int64(common.MaxRequestContentLength))
+	if WebsocketReadDeadline != 0 {
+		conn.SetReadDeadline(time.Now().Add(time.Duration(WebsocketReadDeadline) * time.Second))
+	}
+	if WebsocketWriteDeadline != 0 {
+		conn.SetWriteDeadline(time.Now().Add(time.Duration(WebsocketWriteDeadline) * time.Second))
+	}
+	return NewFuncCodec(conn, conn.WriteJSON, conn.ReadJSON)
 }
 
 // WebsocketHandler returns a handler that serves JSON-RPC to WebSocket connections.
 //
 // allowedOrigins should be a comma-separated list of allowed origin URLs.
 // To allow connections with any origin, pass "*".
-func (srv *Server) WebsocketHandler(allowedOrigins []string) http.Handler {
-	return websocket.Server{
-		Handshake: wsHandshakeValidator(allowedOrigins),
-		Handler: func(conn *websocket.Conn) {
-			if atomic.LoadInt32(&srv.wsConnCount) >= MaxWebsocketConnections {
-				return
-			}
-			atomic.AddInt32(&srv.wsConnCount, 1)
-			wsConnCounter.Inc(1)
-			defer func() {
-				atomic.AddInt32(&srv.wsConnCount, -1)
-				wsConnCounter.Dec(1)
-			}()
-			// Create a custom encode/decode pair to enforce payload size and number encoding
-			conn.MaxPayloadBytes = common.MaxRequestContentLength
-			if WebsocketReadDeadline != 0 {
-				conn.SetReadDeadline(time.Now().Add(time.Duration(WebsocketReadDeadline) * time.Second))
-			}
-			if WebsocketWriteDeadline != 0 {
-				conn.SetWriteDeadline(time.Now().Add(time.Duration(WebsocketWriteDeadline) * time.Second))
-			}
-
-			encoder := func(v interface{}) error {
-				return websocketJSONCodec.Send(conn, v)
-			}
-			decoder := func(v interface{}) error {
-				return websocketJSONCodec.Receive(conn, v)
-			}
-			srv.ServeCodec(NewCodec(conn, encoder, decoder), OptionMethodInvocation|OptionSubscriptions)
-		},
+func (s *Server) WebsocketHandler(allowedOrigins []string) http.Handler {
```

Comment:
do we need to change the variable name from srv to s?

Reply:
oh it's just from ethereum PR. I left it to make it easier to compare to the ethereum PR.
But for the sake of unity with FastWebsockethandler, it is better not to change it.
fixed by [41c7a0d](https://github.com/klaytn/klaytn/pull/1652/commits/41c7a0d2dc08d6d63ddb1591e9d890bf1874c4f5)